### ISSUE TITLE
Overhaul: persistent solvers

### ISSUE THREAD

**Issue by mnwhite** [econ-ark/HARK#97]
As is, a solver object is created for each period, and immediately discarded after reporting its result.  When solving an AgentType with cycles!=1, this is extremely wasteful computationally: much of the "setup" work for each period is done repeatedly, when it only needs to be done *once* for each period in the cycle.

I propose that we allow the solver objects to be persistent and live in a list in the attribute (of the AgentType) solver.  A simple attribute called presolved would indicate whether that solver had already been run at least once; some portions of the solve method would be skipped when presolve=True because they never depend on solution_next.

This would speed up solution of any instance with cycles!=1, but could cause memory problems if T_cycle is large when cycles!=1.  I can't imagine why anyone would do that.  It would also enable easier debugging, as objects created during solution would be available after solving (or after an error).

**Comment by sbenthall** [comment]
Related to #495

**Comment by sbenthall** [comment]
How much setup work is there to the creation of each object?
Is this a serious performance hit?

Can this be fixed with #1211 ?

Wondering if this issue will be fixed or can be closed.

**Comment by mnwhite** [comment]
I'm not sure how much of a performance hit we're taking with it. One way to
test it would be to write code that solves an infinite horizon ConsIndShock
model in as few lines as possible, with none of the HARK overhead. That is,
code that is *only* appropriate for an infinite horizon model and takes
advantage of that fact (by not replicating work across periods). Comparing
the speed of HARK's solver to the "simple solver" would generate an upper
bound on how much could be gained by implementing this idea.

**Comment by sbenthall** [comment]
I believe we've also ran a performance profiler on ConsIndShock which breaks down the performance by function calls.

What I recall we found is that by far the biggest performance cost was in calls to the interpolator.

Such a performance profiling tool could also indicate whether the Solution constructor was causing a big hit.

In general, I would recommend using that profiler when prioritizing changes for performance improvements.
